Objects/regions.c

```diff
@@ -94,9 +120,299 @@ struct regionmetadata {
     regionmetadata* next;
 };
 
-static bool is_bridge_object(PyObject *op);
-static int regionmetadata_has_ancestor(regionmetadata* data, regionmetadata* other);
-static regionmetadata* PyRegion_get_metadata(PyRegionObject* obj);
+static Py_region_ptr_t regionmetadata_merge_root(Py_region_ptr_t self)
+{
+    // Test for local and immutable region
+    if (!HAS_METADATA(self)) {
+        return self;
+    }
+
+    // Return self if it wasn't merged with another region
+    regionmetadata* self_data = REGION_DATA_CAST(self);
+    if (!REGION_PRT_HAS_TAG(self_data->parent, Py_METADATA_MERGE_TAG)) {
+        return self;
+    }
+
+    // FIXME: It can happen that there are several layers in this union-find
+    // structure. It would be efficient to directly update the parent pointers
+    // for deeper nodes.
+    return regionmetadata_merge_root(Py_region_ptr(self_data->parent));
+}
+#define regionmetadata_get_root(self) \
+    regionmetadata_merge_root(REGION_PTR_CAST(self))
+
+static void regionmetadata_open(regionmetadata* self) {
+    assert(HAS_METADATA(self));
+    self->is_open = true;
+}
+
+static bool regionmetadata_is_open(Py_region_ptr_t self) {
+    if (!HAS_METADATA(self)) {
+        return REGION_DATA_CAST(self)->is_open;
+    }
+
+    // The immutable and local region are open by default and can't be closed.
+    return true;
+}
+#define regionmetadata_is_open(self) \
+    regionmetadata_is_open(REGION_PTR_CAST(self))
+
+static void regionmetadata_inc_osc(Py_region_ptr_t self_ptr)
+{
+    if (!HAS_METADATA(self_ptr)) {
+        return;
+    }
+
+    regionmetadata* self = REGION_DATA_CAST(self_ptr);
+    self->osc += 1;
+    regionmetadata_open(self);
+}
+#define regionmetadata_inc_osc(self) \
+    (regionmetadata_inc_osc(REGION_PTR_CAST(self)))
+
+static void regionmetadata_dec_osc(Py_region_ptr_t self_ptr)
+{
+    if (!HAS_METADATA(self_ptr)) {
+        return;
+    }
+
+    REGION_DATA_CAST(self_ptr)->osc -= 1;
+}
+#define regionmetadata_dec_osc(self) \
+    (regionmetadata_dec_osc(REGION_PTR_CAST(self)))
+
+static void regionmetadata_inc_rc(Py_region_ptr_t self)
+{
+    if (HAS_METADATA(self)) {
+        REGION_DATA_CAST(self)->rc += 1;
+    }
+}
+#define regionmetadata_inc_rc(self) \
+    (regionmetadata_inc_rc(REGION_PTR_CAST(self)))
+
+static void regionmetadata_dec_rc(Py_region_ptr_t self_ptr)
+{
+    if (!HAS_METADATA(self_ptr)) {
+        return;
+    }
+
+    // Update RC
+    regionmetadata* self = REGION_DATA_CAST(self_ptr);
+    self->rc -= 1;
+    if (self->rc != 0) {
+        return;
+    }
+
+    // Sort out the funeral by informing everyone about the future freeing
+    Py_XDECREF(self->name);
+    self->name = NULL;
```

Comment:
I wonder if this should be:
```suggestion
    Py_CLEAR(self->name);
```